api/hypershift/v1beta1/nodepool_types.go

```diff
@@ -872,21 +872,51 @@ type AgentNodePoolPlatform struct {
 	AgentLabelSelector *metav1.LabelSelector `json:"agentLabelSelector,omitempty"`
 }
 
+type AzureVMImageType string
+
+const (
+	ImageID          AzureVMImageType = "ImageID"
+	AzureMarketplace AzureVMImageType = "AzureMarketplace"
+)
+
+// AzureVMImage represents the different types of image data that can be provided for an Azure VM.
+// +union
+type AzureVMImage struct {
+	// Type is the type of image data that will be provided to the Azure VM. This can be either "ImageID" or
+	// "MarketplaceImageInfo".
+	//
+	// +kubebuilder:validation:Required
+	// +kubebuilder:validation:Enum:=ImageID;AzureMarketplace
+	// +unionDiscriminator
+	Type AzureVMImageType `json:"azureImageType"`
+
+	// ImageID is the Azure resource ID of a VHD image to use to boot the Azure VMs from.
+	//
+	// +optional
+	// +unionMember, optional
+	ImageID *AzureImageInfo `json:"imageID,omitempty"`
+
+	// MarketplaceImageInfo contains the Azure Marketplace image info to use to boot the Azure VMs from.
+	//
+	// +optional
+	// +unionMember, optional
+	MarketplaceImageInfo *MarketplaceImage `json:"marketplaceImage,omitempty"`
+}
+
 type AzureNodePoolPlatform struct {
 	// VMSize is the Azure VM instance type to use for the nodes being created in the nodepool.
 	//
 	// +kubebuilder:validation:Required
 	// +required
 	VMSize string `json:"vmsize"`
 
-	// ImageID is the id of the image to boot from. If unset, the default image at the location below will be used and
-	// is expected to exist: subscription/<subscriptionID>/resourceGroups/<resourceGroupName>/providers/Microsoft.Compute/images/rhcos.x86_64.vhd.
-	// The <subscriptionID> and the <resourceGroupName> are expected to be the same resource group documented in the
-	// Hosted Cluster specification respectively, hcluster.Spec.Platform.Azure.SubscriptionID and
-	// hcluster.Spec.Platform.Azure.ResourceGroupName.
+	// Image is the image to boot the VMs with
 	//
-	// +optional
-	ImageID string `json:"imageID,omitempty"`
+	// +kubebuilder:validation:XValidation:rule="self == oldSelf", message="Image is immutable"
+	// +kubebuilder:validation:Required
+	// +immutable
```

Comment:
+immutable was a placeholder before CEL supported immutability. It has no effect, can we drop it?